com.unity.ml-agents.extensions/Runtime/Sensors/GridSensorComponent.cs

```diff
@@ -202,68 +184,149 @@ public int ObservationStacks
+            m_BoxOverlapChecker = new BoxOverlapChecker(
                 m_CellScale,
                 m_GridSize,
                 m_RotateWithAgent,
-                m_ChannelDepths,
-                m_DetectableObjects,
                 m_ColliderMask,
-                m_DepthType,
                 gameObject,
-                m_CompressionType,
-                m_MaxColliderBufferSize,
-                m_InitialColliderBufferSize
+                m_DetectableTags,
+                m_InitialColliderBufferSize,
+                m_MaxColliderBufferSize
             );
 
-            if (ObservationStacks != 1)
+            // debug data is positive int value and will trigger data validation exception if SensorCompressionType is not None.
+            m_DebugSensor = new GridSensorBase("DebugGridSensor", m_CellScale, m_GridSize, m_DetectableTags, SensorCompressionType.None);
+            m_BoxOverlapChecker.RegisterDebugSensor(m_DebugSensor);
+
+            if (m_UseOneHotTag)
             {
-                return new ISensor[] { new StackingSensor(m_Sensor, ObservationStacks) };
+                var sensor = new OneHotGridSensor(m_SensorName + "-OneHot", m_CellScale, m_GridSize, m_DetectableTags, m_CompressionType);
+                if (ObservationStacks != 1)
+                {
+                    m_Sensors.Add(new StackingSensor(sensor, ObservationStacks));
+                }
+                else
+                {
+                    m_Sensors.Add(sensor);
+                }
+                m_BoxOverlapChecker.RegisterSensor(sensor);
             }
-            return new ISensor[] { m_Sensor };
+            if (m_CountColliders)
+            {
+                var sensor = new CountingGridSensor(m_SensorName + "-Counting", m_CellScale, m_GridSize, m_DetectableTags, m_CompressionType);
+                if (ObservationStacks != 1)
+                {
+                    m_Sensors.Add(new StackingSensor(sensor, ObservationStacks));
+                }
+                else
+                {
+                    m_Sensors.Add(sensor);
+                }
+                m_BoxOverlapChecker.RegisterSensor(sensor);
+            }
+            var customSensors = GetGridSensors();
+            if (customSensors != null)
+            {
+                foreach (var sensor in customSensors)
+                {
+                    if (ObservationStacks != 1)
+                    {
+                        m_Sensors.Add(new StackingSensor(sensor, ObservationStacks));
+                    }
+                    else
+                    {
+                        m_Sensors.Add(sensor);
+                    }
+                    m_BoxOverlapChecker.RegisterSensor(sensor);
+                }
+            }
+            // Only one sensor needs to reference the boxOverlapChecker, so that it gets updated exactly once
+            ((GridSensorBase)m_Sensors[0]).m_BoxOverlapChecker = m_BoxOverlapChecker;
```

Comment:
m_Sensors will be empty if both m_UseOneHotTag and m_CountColliders are false, right? I think you need to check m_Sensors.Count first.

Reply:
Fixed it. Also added tests for this.